### PR TITLE
Use custom exceptions

### LINKED ISSUE #86: Tube#recvuntil should raise TimeoutError when time limit exceeded

### PR DESCRIPTION
**Pull request by hanhanW**
This change is [Reviewable](https://reviewable.io/reviews/peter50216/pwntools-ruby/87)

### CODE REVIEW

**Comment by hanhanW** [comment]
This also address issue #86

**Comment by david942j** [comment]
I think in these cases, they are worth to have a custom error, just change to a more specific name.
Maybe change EOFError to EndOfTubeError and NameError to ConstantNotFoundError?

File: lib/pwnlib/constants/constants.rb [hunk 1]
```diff
@@ -36,17 +37,23 @@ def respond_to_missing?(method, _include_all)
       # @return [Constant]
       #   The evaluated result.
       #
+      # @raise [Pwnlib::Errors::ConstantNotFoundError]
+      #   If no value provided for variables.
```
Comment (david942j) [inline lib/pwnlib/constants/constants.rb]: change this description

File: lib/pwnlib/constants/constants.rb [hunk 2]
```diff
@@ -36,17 +37,23 @@ def respond_to_missing?(method, _include_all)
...
+      # @example
+      #   eval('meow')
+      #   # Pwnlib::Errors::ConstantNotFoundError: no value provided for variables: meow
       def eval(str)
         return str unless str.instance_of?(String)
         begin
           val = calculator.evaluate!(str.strip).to_i
         rescue Dentaku::UnboundVariableError => e
-          raise NameError, e.message
+          raise ::Pwnlib::Errors::ConstantNotFoundError, e.message
```
Comment (david942j) [inline lib/pwnlib/constants/constants.rb]: use `e. unbound_variables` to get undefined variables, and use a more meaningful description
`"Constants are not defined: #{e.unbound_variables.join(', ')}"`

File: lib/pwnlib/constants/constants.rb (hunk 1 above)
Comment (david942j) [inline lib/pwnlib/constants/constants.rb]: and change this description as well

File: test/tubes/tube_test.rb [hunk 3]
```diff
@@ -60,7 +60,7 @@ def basic_tube
 
     class << t
       def recv_raw(_n)
-        raise EOFError if io.eof?
+        raise Pwnlib::Errors::EndOfTubeError if io.eof?
```
Comment (david942j) [inline test/tubes/tube_test.rb]: `::Pwnlib`

File: test/tubes/tube_test.rb [hunk 4]
```diff
@@ -123,11 +123,18 @@ def test_recvuntil
     assert_equal('worldHello, ', t.recvuntil(' '))
     t.unrecv('Hello, world!')
     assert_equal('Hello,', t.recvuntil(' wor', drop: true))
-    assert_equal('', t.recvuntil('DARKHH', drop: true, timeout: 0.1))
+
+    # test timeout
+    assert_raises(::Pwnlib::Errors::TimeoutError) do
+      assert_equal('', t.recvuntil('DARKHH', drop: true, timeout: 0.1))
```
Comment (david942j) [inline test/tubes/tube_test.rb]: remove `assert_equal`

File: test/tubes/tube_test.rb [hunk 5]
```diff
@@ -123,11 +123,18 @@ def test_recvuntil
...
+    end
 
     t = basic_tube
     t.unrecv('meow')
-    assert_equal('', t.recvuntil('DARKHH'))
+    # test EOF
+    assert_raises(::Pwnlib::Errors::EndOfTubeError) do
+      assert_equal('', t.recvuntil('DARKHH'))
```
Comment (david942j) [inline test/tubes/tube_test.rb]: remove `assert_equal`

File: test/tubes/tube_test.rb [hunk 6]
```diff
@@ -145,7 +152,10 @@ def test_recvline
 
     t = basic_tube
     t.unrecv('Hello, world')
-    assert_equal('', t.recvline)
+    # test EOF
+    assert_raises(::Pwnlib::Errors::EndOfTubeError) do
+      assert_equal('', t.recvline)
```
Comment (david942j) [inline test/tubes/tube_test.rb]: remove `assert_equal`

File: test/tubes/tube_test.rb [hunk 7]
```diff
@@ -168,10 +178,17 @@ def test_recvpred
     r = /H.*w/
     10.times { assert_match(r, t.recvpred { |data| data =~ r }) }
     r = /H.*W/
-    assert_match('', t.recvpred(timeout: 0.01) { |data| data =~ r })
+
+    # test timeout
+    assert_raises(::Pwnlib::Errors::TimeoutError) do
+      assert_match('', t.recvpred(timeout: 0.01) { |data| data =~ r })
```
Comment (david942j) [inline test/tubes/tube_test.rb]: remove `assert_match`

File: test/tubes/tube_test.rb [hunk 8]
```diff
@@ -168,10 +178,17 @@ def test_recvpred
...
+    end
     t = basic_tube
     t.unrecv('darkhh')
-    assert_match('', t.recvpred { |data| data =~ r })
+    # test EOF
+    assert_raises(::Pwnlib::Errors::EndOfTubeError) do
+      assert_match('', t.recvpred { |data| data =~ r })
```
Comment (david942j) [inline test/tubes/tube_test.rb]: remove `assert_match`

File: lib/pwnlib/constants/constants.rb [hunk 9]
```diff
@@ -36,17 +37,23 @@ def respond_to_missing?(method, _include_all)
...
+      #   If no value provided for variables.
+      #
       # @example
       #   eval('O_CREAT')
       #   #=> Constant('(O_CREAT)', 0x40)
       #   eval('O_CREAT | O_APPEND')
       #   #=> Constant('(O_CREAT | O_APPEND)', 0x440)
+      # @example
+      #   eval('meow')
+      #   # Pwnlib::Errors::ConstantNotFoundError: no value provided for variables: meow
```
Comment (david942j) [inline lib/pwnlib/constants/constants.rb]: and don't forget here if the error message is changed

**Pull request merged**